### PR TITLE
Refs #5020 Static data now allows different sizes between tests. [5029]

### LINKED ISSUE #463: Fast-RTPS latency performance test error for large data [5020]
When I use the latency test of the Fast-RTPS performance test with version 1.7.0, i see the following error on the subscriber side on Linux x86_64 system.

Terminal-1
------------------
```
fastrtps/test/performance/LatencyTest publisher -l
Performing test with 1 subscribers and 10000 samples
Data Sub Matched 
Data Pub Matched 
Command Pub Matched 
Command Sub Matched 
Overhead 0.0326344 ns
DISCOVERY COMPLETE 
Printing round-trip times in us, statistics for 10000 samples
   Bytes, Samples,   stdev,    mean,     min,     50%,     90%,     99%,  99.99%,     max
--------,--------,--------,--------,--------,--------,--------,--------,--------,--------,
```

Terminal-2
----------------
```
fastrtps/test/performance/LatencyTest subscriber
DISCOVERY COMPLETE 
Preparing test with data size: 16
RCOMMAND: 1
Publisher has new test ready...
Testing with data size: 16
2019-03-25 17:32:47.106 [RTPS_HISTORY Error] Change payload size of '64004' bytes is larger than the history payload size of '17003' bytes and cannot be resized. -> Function add_change
2019-03-25 17:32:48.106 [RTPS_HISTORY Error] Change payload size of '64004' bytes is larger than the history payload size of '17003' bytes and cannot be resized. -> Function add_change
2019-03-25 17:32:49.106 [RTPS_HISTORY Error] Change payload size of '64004' bytes is larger than the history payload size of '17003' bytes and cannot be resized. -> Function add_change
2019-03-25 17:32:50.106 [RTPS_HISTORY Error] Change payload size of '64004' bytes is larger than the history payload size of '17003' bytes and cannot be resized. -> Function add_change
2019-03-25 17:32:51.106 [RTPS_HISTORY Error] Change payload size of '64004' bytes is larger than the history payload size of '17003' bytes and cannot be resized. -> Function add_change
2019-03-25 17:32:52.106 [RTPS_HISTORY Error] Change payload size of '64004' bytes is larger than the history payload size of '17003' bytes and cannot be resized. -> Function add_change
```

How do I mitigate this issue?

Thanks
Anup

### PR DESCRIPTION
**Pull request by LuisGP**
Fixes #463
Allows multiple static data sizes between tests.
Help messages fixed.

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1415/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1384/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1395/)

File: test/performance/ThroughputSubscriber.cpp [hunk 1]
```diff
@@ -468,20 +489,29 @@ ThroughputSubscriber::ThroughputSubscriber(bool reliable, uint32_t pid, bool hos
...
-        Domain::removeSubscriber(mp_datasub);
-        Domain::unregisterType(mp_par, "ThroughputType"); // Unregister as we will register it later with correct size
     }
+    else
+    {
+        delete throughput_t;
+        throughput_t = nullptr;
+    }
+    //if (mp_datasub != nullptr)
+    //{
```
Comment (richiware) [inline test/performance/ThroughputSubscriber.cpp]: A commented block of code¿?

File: test/performance/ThroughputSubscriber.cpp [hunk 2]
```diff
@@ -158,13 +158,18 @@ void ThroughputSubscriber::CommandSubListener::onSubscriptionMatched(Subscriber*
     m_up.disc_cond_.notify_one();
 }
 
-void ThroughputSubscriber::CommandSubListener::onNewDataMessage(Subscriber* subscriber)
+void ThroughputSubscriber::CommandSubListener::onNewDataMessage(Subscriber*)
+{
+    m_up.stop_cond_.notify_one();
```
[the remark this answers is not in the document]
Comment (LuisGP) [reply test/performance/ThroughputSubscriber.cpp]: In this case, that *condition_variable* wasn't being used, so I removed.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1551/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1514/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1525/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1554/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1517/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1528/)

**Comment by MiguelCompany** [comment]
@LuisGP Please rebase to resolve the conflicts

File: test/performance/ThroughputPublisher.cpp [hunk 3]
```diff
@@ -283,18 +274,10 @@ ThroughputPublisher::ThroughputPublisher(bool reliable, uint32_t pid, bool hostn
     t_overhead_ = std::chrono::duration<double, std::micro>(t_end_ - t_start_) / 1001;
     std::cout << "Overhead " << t_overhead_.count() << " us"  << std::endl;
 
-    if (mp_datapub == nullptr || mp_commandsub == nullptr || mp_commandpub == nullptr)
+    if (/*mp_datapub == nullptr || */mp_commandsub == nullptr || mp_commandpub == nullptr)
```
Comment (MiguelCompany) [inline test/performance/ThroughputPublisher.cpp]: I think the commented code should just be removed.

File: test/performance/ThroughputPublisher.cpp [hunk 4]
```diff
@@ -493,14 +479,20 @@ bool ThroughputPublisher::test(uint32_t test_time, uint32_t recovery_time_ms, ui
     {
         DynamicTypeBuilderFactory::delete_instance();
         DynamicDataFactory::get_instance()->delete_data(m_DynData);
-        pubAttr = mp_datapub->getAttributes();
-        Domain::removePublisher(mp_datapub);
-        Domain::unregisterType(mp_par, "ThroughputType");
     }
     else
     {
+        //delete(latency_t);
```
Comment (MiguelCompany) [inline test/performance/ThroughputPublisher.cpp]: Remove this code

File: test/performance/ThroughputSubscriber.cpp [hunk 5]
```diff
@@ -190,85 +194,91 @@ void ThroughputSubscriber::CommandSubListener::onNewDataMessage(Subscriber* subs
...
-
-                    m_up.m_DynData = DynamicDataFactory::get_instance()->create_data(m_up.m_pDynType);
+                    m_DynData = DynamicDataFactory::get_instance()->create_data(m_pDynType);
                 }
                 else
                 {
-                    delete(m_up.throughputin);
-                    //m_up.throughputin = nullptr;
-                    m_up.throughputin = new ThroughputType((uint16_t)m_up.m_datasize);
+                    delete(throughput_t);
```
Comment (MiguelCompany) [inline test/performance/ThroughputSubscriber.cpp]: Shouldn't we check nullptr before deleting?

File: test/performance/ThroughputSubscriber.cpp [hunk 6]
```diff
@@ -469,20 +477,12 @@ ThroughputSubscriber::ThroughputSubscriber(bool reliable, uint32_t pid, bool hos
     t_overhead_ = std::chrono::duration<double, std::micro>(t_end_ - t_start_) / 1001;
     std::cout << "Overhead " << t_overhead_.count() << std::endl;
 
-    if (mp_datasub == nullptr || mp_commandsub == nullptr || mp_commandpubli == nullptr)
+    if (/*mp_datasub == nullptr || */mp_commandsub == nullptr || mp_commandpubli == nullptr)
```
Comment (MiguelCompany) [inline test/performance/ThroughputSubscriber.cpp]: Commented code should be removed

File: test/performance/ThroughputSubscriber.cpp (hunk 5 above)
Comment (LuisGP) [reply test/performance/ThroughputSubscriber.cpp]: I think isn't necessary:

`delete(nullptr)` is defined to do nothing. About potential double free: In the constructor is set to `nullptr`. In line 555 is set to `nullptr` after deleting it. So the only deletion, if I'm not wrong, is performed in that line is when reserved at line 214.
So double free shouldn't happen.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1685/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1645/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1657/)

**Comment by richiware** [comment]
Merged. We expect to be run by release/1.8.0 PR.

**Pull request merged**